File: Python3/프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    answer = 0
    walk_set = set((0, 0, 0, 0))
    drc = {'U': [1, 0], 'D': [-1, 0], 'L': [0, -1], 'R': [0, 1]}
    answer = 0
    r, c, dr, dc, nr, nc = 0, 0, 0, 0, 0, 0
    
    for dir in dirs:
        dr, dc = drc[dir]
        nr, nc = r + dr, c + dc
        if abs(nr) <= 5 and abs(nc) <= 5:
            if (r, c, nr, nc) not in walk_set and (nr, nc, r, c) not in walk_set:
                answer += 1
                walk_set.add((r, c, nr, nc))
            r, c = nr, nc
            # print("(", nr, ",", nc, ")로 이동...")
    
    return answer

File: C++/프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <vector>
#include <unordered_map>
using namespace std;

int solution(string dirs) {
    bool visited[11][11][11][11] = {0};
    int r = 5, c = 5, nr = 0, nc = 0;
    unordered_map<char, vector<int>> drc;
    drc['U'] = {1, 0};
    drc['D'] = {-1, 0};
    drc['L'] = {0, -1};
    drc['R'] = {0, 1};
    
    int answer = 0;
    
    for(auto& dir : dirs) {
        nr = r + drc[dir][0];
        nc = c + drc[dir][1];
        
        if(0 <= nr and nr < 11 and 0 <= nc and nc < 11){
            if(!visited[r][c][nr][nc]) {
                visited[r][c][nr][nc] = true;
                visited[nr][nc][r][c] = true;
                answer++;
            }
            r = nr;
            c = nc;
        }
    }
    
    return answer;
}